src/libraries/System.Security.Cryptography/tests/X509Certificates/X509FilesystemTests.Windows.cs

```diff
@@ -0,0 +1,464 @@
+        {
+            Pkcs12SafeContents keys = new Pkcs12SafeContents();
+            Pkcs12SafeContents certs = new Pkcs12SafeContents();
+            DateTimeOffset notBefore = DateTimeOffset.UtcNow.AddMinutes(-5);
+            DateTimeOffset notAfter = notBefore.AddMinutes(10);
+
+            // Every other key is a machine key. Depending on the minute when this line of code runs, it's
+            // either starting with the first, or with the second.
+            int machineKeySkew = DateTime.UtcNow.Minute % 2;
+
+            PbeParameters pbeParams = new PbeParameters(
+                PbeEncryptionAlgorithm.TripleDes3KeyPkcs12,
+                HashAlgorithmName.SHA1,
+                1);
+
+            for (int i = 0; i < certAndKeyCount; i++)
+            {
+                using (RSA key = RSA.Create(1024))
+                {
+                    CertificateRequest req = new CertificateRequest(
+                        $"CN={name}.{i}",
+                        key,
+                        HashAlgorithmName.SHA256,
+                        RSASignaturePadding.Pkcs1);
+
+                    using (X509Certificate2 cert = req.CreateSelfSigned(notBefore, notAfter))
+                    {
+                        Pkcs12ShroudedKeyBag keyBag = keys.AddShroudedKey(key, "", pbeParams);
+
+                        if (capi)
+                        {
+                            keyBag.Attributes.Add(s_capiCsp);
+                        }
+
+                        if (int.IsEvenInteger(i + machineKeySkew))
+                        {
+                            keyBag.Attributes.Add(s_machineKey);
+                        }
+
+                        certs.AddCertificate(cert);
+                    }
+                }
+            }
+
+            for (int i = 0; i < certOnlyCount; i++)
+            {
+                using (RSA key = RSA.Create(1024))
+                {
+                    CertificateRequest req = new CertificateRequest(
+                        $"CN={name}.co.{i}",
+                        key,
+                        HashAlgorithmName.SHA256,
+                        RSASignaturePadding.Pkcs1);
+
+                    using (X509Certificate2 cert = req.CreateSelfSigned(notBefore, notAfter))
+                    {
+                        certs.AddCertificate(cert);
+                    }
+                }
+            }
+
+            Pkcs12Builder builder = new Pkcs12Builder();
+            builder.AddSafeContentsEncrypted(certs, "", pbeParams);
+            builder.AddSafeContentsUnencrypted(keys);
+            builder.SealWithMac("", HashAlgorithmName.SHA1, 1);
+            return builder.Encode();
+        }
+
+        private static void EnsureNoKeysGained<TState>(TState state, Func<TState, IDisposable> importer)
+        {
+            const int ERROR_ACCESS_DENIED = (unchecked((int)0x80010005));
+
+            // In the good old days, before we had threads or parallel processes, these tests would be easy:
```

Comment:
Just asking if you considered it, but isn't the on-disk name of the key basically a combo of SID and UniqueName, at least for CNG? Can we make these tests more deterministic or reliable by looking for keys by name, not just diff-counts?

Reply:
Ah, yeah, that's fair.